File: scripts/eval/ci_compare.py
#!/usr/bin/env python3
import json
import sys
from pathlib import Path


def format_diff(current, previous, invert_color=False):
    if current is None or previous is None:
        return ""
    diff = current - previous
    if abs(diff) < 0.01:
        return ""

    sign = "+" if diff > 0 else ""
    val_str = f"{sign}{diff:.2f}" if isinstance(diff, float) else f"{sign}{int(diff)}"

    # We use simple github markdown, no rich formatting needed, no BBCode
    return f"({val_str})"


def score_emoji(score):
    if score is None:
        return "❓"
    return ["", "🔴", "🟠", "🟡", "🟢", "✅"][round(score)]


def main():
    if len(sys.argv) != 4:
        print("Usage: ci_compare.py <main_json> <pr_json> <output_md>")
        sys.exit(1)

    main_path = Path(sys.argv[1])
    pr_path = Path(sys.argv[2])
    out_path = Path(sys.argv[3])

    if not main_path.exists():
        print(f"Main json not found at {main_path}, proceeding blindly.")
        main_history = {}
    else:
        main_history = json.loads(main_path.read_text())

    if not pr_path.exists():
        print(f"PR json not found at {pr_path}. Fail.")
        sys.exit(1)

    pr_history = json.loads(pr_path.read_text())

    pr_scores = [
        v.get("avg_score")
        for v in pr_history.values()
        if v.get("avg_score") is not None
    ]
    pr_total = sum(pr_scores) / len(pr_scores) if pr_scores else 0

    main_scores = [
        v.get("avg_score")
        for v in main_history.values()
        if v.get("avg_score") is not None
    ]
    main_total = sum(main_scores) / len(main_scores) if main_scores else 0
    total_diff = format_diff(pr_total, main_total)

    lines = [
        "## Agent Prompt Evaluation",
        f"**Overall Score:** {score_emoji(pr_total)} {pr_total:.2f}/5 {total_diff}",
        "",
    ]

    if pr_total < 4.0:
        lines.append("> 🔴 **FAILED:** Overall score is below the 4.0 threshold.")
        lines.append("")
    elif pr_total < main_total:
        lines.append(
            f"> 🟠 **WARNING:** Quality regression detected! Score dropped from {main_total:.2f} to {pr_total:.2f}."
        )
        lines.append("")

    lines.append("| Case | Score | Token Score |")
    lines.append("|------|-------|-------------|")

    for case_id, pr_data in pr_history.items():
        main_data = main_history.get(case_id, {})

        pr_score = pr_data.get("avg_score")
        main_score = main_data.get("avg_score")
        score_df = format_diff(pr_score, main_score)

        pr_token_score = pr_data.get("token_score")
        main_token_score = main_data.get("token_score")
        token_score_df = format_diff(
            pr_token_score, main_token_score, invert_color=True
        )

        sc_str = (
            f"{score_emoji(pr_score)} {pr_score:.2f} {score_df}".strip()
            if pr_score is not None
            else "ERR"
        )
        ts_str = f"{pr_token_score or '?'} {token_score_df}".strip()

        lines.append(f"| {case_id} | {sc_str} | {ts_str} |")

    out_path.write_text("\n".join(lines))

    # Determine exit status
    if pr_total < 4.0:
        print("Score below 4.0 threshold.")
        sys.exit(1)

    print("Evaluation checks passed.")
    sys.exit(0)


if __name__ == "__main__":
    main()

File: cmd/agents-cli/cmd/ci_compare.go
package cmd

import (
	"encoding/json"
	"fmt"
	"math"
	"os"
	"strings"

	"github.com/spf13/cobra"
)

func formatDiff(current, previous *float64, invertColor bool) string {
	if current == nil || previous == nil {
		return ""
	}
	diff := *current - *previous
	if math.Abs(diff) < 0.01 {
		return ""
	}

	sign := "+"
	if diff < 0 {
		sign = ""
	}
	valStr := fmt.Sprintf("%s%.2f", sign, diff)
	// if it is an integer, drop .00
	if diff == math.Trunc(diff) {
		valStr = fmt.Sprintf("%s%d", sign, int(diff))
	}

	return fmt.Sprintf("(%s)", valStr)
}

func scoreEmoji(score *float64) string {
	if score == nil {
		return "❓"
	}
	emojis := []string{"", "🔴", "🟠", "🟡", "🟢", "✅"}
	idx := int(math.Round(*score))
	if idx < 0 {
		idx = 0
	}
	if idx > 5 {
		idx = 5
	}
	return emojis[idx]
}

type evalData struct {
	AvgScore   *float64 `json:"avg_score"`
	TokenScore *float64 `json:"token_score"`
}

var ciCompareCmd = &cobra.Command{
	Use:   "ci-compare [MAIN_JSON] [PR_JSON] [OUTPUT_MD]",
	Short: "Compare evaluation results for CI",
	Args:  cobra.ExactArgs(3),
	Run: func(cmd *cobra.Command, args []string) {
		mainPath := args[0]
		prPath := args[1]
		outPath := args[2]

		mainHistory := make(map[string]evalData)
		if b, err := os.ReadFile(mainPath); err != nil {
			fmt.Printf("Main json not found at %s, proceeding blindly.\n", mainPath)
		} else {
			json.Unmarshal(b, &mainHistory)
		}

		prHistory := make(map[string]evalData)
		if b, err := os.ReadFile(prPath); err != nil {
			fmt.Printf("PR json not found at %s. Fail.\n", prPath)
			os.Exit(1)
		} else {
			json.Unmarshal(b, &prHistory)
		}

		var prScores []float64
		for _, v := range prHistory {
			if v.AvgScore != nil {
				prScores = append(prScores, *v.AvgScore)
			}
		}
		var prTotal float64 = 0
		if len(prScores) > 0 {
			var sum float64 = 0
			for _, s := range prScores {
				sum += s
			}
			prTotal = sum / float64(len(prScores))
		}

		var mainScores []float64
		for _, v := range mainHistory {
			if v.AvgScore != nil {
				mainScores = append(mainScores, *v.AvgScore)
			}
		}
		var mainTotal float64 = 0
		if len(mainScores) > 0 {
			var sum float64 = 0
			for _, s := range mainScores {
				sum += s
			}
			mainTotal = sum / float64(len(mainScores))
		}

		totalDiff := formatDiff(&prTotal, &mainTotal, false)

		lines := []string{
			"## Agent Prompt Evaluation",
			fmt.Sprintf("**Overall Score:** %s %.2f/5 %s", scoreEmoji(&prTotal), prTotal, totalDiff),
			"",
		}

		if prTotal < 4.0 {
			lines = append(lines, "> 🔴 **FAILED:** Overall score is below the 4.0 threshold.", "")
		} else if prTotal < mainTotal {
			lines = append(lines, fmt.Sprintf("> 🟠 **WARNING:** Quality regression detected! Score dropped from %.2f to %.2f.", mainTotal, prTotal), "")
		}

		lines = append(lines, "| Case | Score | Token Score |")
		lines = append(lines, "|------|-------|-------------|")

		for caseID, prData := range prHistory {
			mainData := mainHistory[caseID]

			scoreDf := formatDiff(prData.AvgScore, mainData.AvgScore, false)
			tokenScoreDf := formatDiff(prData.TokenScore, mainData.TokenScore, true)

			scStr := "ERR"
			if prData.AvgScore != nil {
				scStr = strings.TrimSpace(fmt.Sprintf("%s %.2f %s", scoreEmoji(prData.AvgScore), *prData.AvgScore, scoreDf))
			}

			tsVal := "?"
			if prData.TokenScore != nil {
				tsVal = fmt.Sprintf("%v", *prData.TokenScore)
			}
			tsStr := strings.TrimSpace(fmt.Sprintf("%s %s", tsVal, tokenScoreDf))

			lines = append(lines, fmt.Sprintf("| %s | %s | %s |", caseID, scStr, tsStr))
		}

		os.WriteFile(outPath, []byte(strings.Join(lines, "\n")), 0644)

		if prTotal < 4.0 {
			fmt.Println("Score below 4.0 threshold.")
			os.Exit(1)
		}

		fmt.Println("Evaluation checks passed.")
	},
}

func init() {
	rootCmd.AddCommand(ciCompareCmd)
}
